### ISSUE TITLE
WASM32-WASI NN support

### ISSUE THREAD

**Issue by profitgrowinginnovator** [tracel-ai/burn#2520]
### Feature description

There are several requests for WASM (webworkers,...). Please add WASM32-WASI with NN WIT support

### Feature motivation

Solutions like WASMEdge allow for LLM inference (https://wasmedge.org/docs/develop/rust/wasinn/llm_inference/), so it would be excellent if any Burn LLM model could be compiled into a WASM32-WASI with NN WIT support and ran in serverless, edge and other constrained environments.  

### (Optional) Suggest a Solution

A wasi-nn feature allows a cargo build to target wasm32-wasip2 and NN WIT

**Comment by antimora** [comment]
Yes, I think it'd be useful. 

So basically it would mean to implement a new backend that can use wasi-nn apis and have the code compiled to wasm32?

Linking wasi-nn https://github.com/WebAssembly/wasi-nn/tree/main

**Comment by profitgrowinginnovator** [comment]
Yes you should be able to add the target wasm32-wasip1/2 and get a wasm which you can run for instance on wasmedge

**Comment by profitgrowinginnovator** [comment]
Great question. Wasm32-unknown-unknown is mainly used inside browsers where access to file systems is very restricted. Wasi allows the use of WIT, standardised interfaces. Some are for IO, key-value, cryptography and one is especially interesting for AI, the wasi-nn [neural networks]. This interface allows access to ML accelerators, e.g. OpenVINO, CUDA,... which can make the code run exponentially faster. WASI is made for device and cloud deployments where the unknown one is mainly for browsers. Hopefully that makes sense now. You might want to support wasip2 directly although nn is still pending.

**Comment by nathanielsimard** [comment]
We're unlikely to create another backend using an external execution engine, as most of our efforts are focused on developing our own optimizations, kernels, and compiler tools. If there is an API to support GPU execution (SPIRV, CUDA, Metal, etc.) then we could target those in CubeCL.

**Comment by antimora** [comment]
It would worth coming back to this when `wasi-nn` matures and becomes generally available on major cloud services. I do see benefits deploying models to the edge compute easily. However, it's possible that the cloud services might just expose GPU execution (SPIRV, CUDA, Metal, etc.) APIs directly instead of having a new API layer such as `wasi-nn`. In which case, CubeCL will be more useful.

**Comment by diversable** [comment]
I think you might be under-estimating the potential of `wasi` & `wasi-nn` -- I understand waiting until it matures, but having native support for deploying to `wasi` / `wasi-nn` will make Burn _much_ more accessible on alternative & specialized cloud platforms, as well as in K8s (which now has the capability to run `wasi` programs as well as full-fat containers). 

One of the most important parts of the deploy-to-wasi-platform story is that `wasi` programs start up in around ~10ms, whereas containers take a minimum of ~300ms (and more often 1-2 seconds). 

Even now, platforms like Fermyon's Spin (cloud platform & the SpinKube K8s platform), WasmCloud, WasmEdge, etc. offer much more attractive deployment targets for latency-sensitive applications like recommendations & personalization systems for eCommerce, which obviously has huge market potential. 

So, please keep a close eye on the development of `wasi` and `wasi-nn` and start implementing support as soon as is feasible.

**Comment by antimora** [comment]
Outdated and not planned
